File: 백준/Gold/2623. 음악프로그램/음악프로그램.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            adj.add(new ArrayList<>());
        }

        int[] ind = new int[n + 1];

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int[] nums = new int[st.countTokens()];
            for (int j = 0; st.hasMoreTokens(); j++) {
                nums[j] = Integer.parseInt(st.nextToken());
            }
            for (int j = 1; j < nums.length - 1; j++) {
                adj.get(nums[j]).add(nums[j + 1]);
                ind[nums[j + 1]]++;
            }
        }

        List<Integer> candidates = new ArrayList<>();
        for (int i = 1; i <= n; i++) {
            if (ind[i] == 0) {
                candidates.add(i);
            }
        }

        List<Integer> result = new ArrayList<>();
        while (!candidates.isEmpty()) {
            int cur = candidates.remove(candidates.size() - 1);
            result.add(cur);
            for (int nxt : adj.get(cur)) {
                ind[nxt]--;
                if (ind[nxt] == 0) {
                    candidates.add(nxt);
                }
            }
        }

        if (result.size() != n) {
            System.out.println(0);
        } else {
            for (int num : result) {
                System.out.println(num);
            }
        }
    }
}

File: 백준/Gold/2623. 음악프로그램/음악프로그램.py
# https://www.acmicpc.net/problem/2623
n, m = map(int, input().split())
adj = [[] for _ in range(n + 1)]
ind = [0 for _ in range(n + 1)]

for _ in range(m):
    nums = list(map(int, input().split()))
    for i in range(1, nums[0]):
        # 그래프 생성 - 방향성 존재
        adj[nums[i]].append(nums[i + 1])
        # 진입하는 경우 in-degree 증가
        ind[nums[i + 1]] += 1

# in-degree = 0 후보군 넣기
candidates = []
for i in range(1, n + 1):
    if ind[i] == 0:
        candidates.append(i)

result = []
while candidates:
    # 후보군에서 추출
    cur = candidates.pop()
    result.append(cur)
    # 추출 후 연결된 다음 노드들의 in-degree 제거
    for nxt in adj[cur]:
        ind[nxt] -= 1
        if ind[nxt] == 0:
            candidates.append(nxt)

if len(result) != n:
    print(0)
else:
    for num in result:
        print(num)